src/kaleidoscope/hid.cpp

```diff
@@ -150,11 +221,38 @@ uint8_t getKeyboardLEDs() {
 
 
 void sendKeyboardReport() {
+
+  // Before sending the report, we add any modifier flags that are currently allowed,
+  // based on the latest keypress:
+  pressModFlags(current_mod_flags & mod_flags_allowed);
+
+  // If a key has just toggled on, we assume that the user expects that key's keycode to
+  // be sent as a new event immediately, so we make sure that a report is sent without
+  // that keycode. In the usual case, this won't result in any difference from the
+  // previous report, so no extra report will be sent. If there is a difference in the
+  // modifiers byte, an extra report would be sent later, regardless. The only time we
+  // should genuinely see an extra report from this is if we need one because a user
+  // rolled over from `X` to `x`, or if they've got a keymap with two `e` keys because
+  // they find it problematic to double-tap the same key frequently.
   WITH_BOOTKEYBOARD_PROTOCOL {
+    if (toggled_on_keycode) {
+      BootKeyboard.release(toggled_on_keycode);
```

Comment:
Is the "release" behavior what you really want here or do you just want to force another key report to be sent before sending the new report?

Reading this bit of code, I'm pretty sure we're going to run into issues if multiple keys are changed in a single scan cycle.

Reply:
This (generally) gets called once per scan cycle, so if two keys toggle on in the same cycle, the second one's keycode will clobber the first one's.

I'm sure we do want to call `release()` here — this is for the case where a user rolls over from a mod-flagged key to a non-flagged key, but both have the same keycode. Without sending an extra report with that keycode removed, the OS won't treat that as a separate keypress, and to the user, it will seem like the second key just failed to register.

Reply:
Ok. The explanation of why we want to call release there is definitely something we should migrate into code comments. It's one of those very subtle problems that isn't at all obvious from inspecting the code.

Reply:
Heh. I thought I had done so, in the last sentence of the preceding comment block (lines 233-236). I'll see if I can clarify it more.